feat: trigger MultichainAccountService basic functionality updates

fabiobozzo opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

Add multichain account provider state management when basic functionality toggle changes on mobile.

When users toggle the "Basic Functionality" setting on mobile, account providers need to be managed appropriately - non-EVM providers should be disabled when basic functionality is OFF, and all providers should be enabled with wallet alignment when basic functionality is ON.

## **Changelog**

CHANGELOG entry: null

This PR integrates with the `MultichainAccountService.setBasicFunctionality()` method to:
- Update provider disabled states based on basic functionality preference
- Trigger multichain wallet alignment when basic functionality is enabled

1. **BasicFunctionalityModal** - Main settings toggle modal
2. **ConfirmTurnOnBackupAndSyncModal** - Backup & sync confirmation flow

Both now call `Engine.context.MultichainAccountService.setBasicFunctionality({ enabled })` after updating Redux state.

## **Related issues**

[MUL-343](https://consensyssoftware.atlassian.net/browse/MUL-343)

Relates to: https://github.com/MetaMask/metamask-extension/pull/35190, https://github.com/MetaMask/core/pull/6332

## **Manual testing steps**

Scenario: user enables basic functionality from settings
Given basic functionality is currently disabled
And user is in Settings > Security & Privacy > Basic Functionality

When user toggles "Basic Functionality" to ON
Then Redux state updates to basicFunctionalityEnabled: true
And MultichainAccountService.setBasicFunctionality is called with enabled: true
And console shows "MultichainAccountService: Setting basic functionality enabled"
And console shows "Triggered wallet alignment..."
And all multichain account providers are enabled

---

Scenario: user disables basic functionality from settings  
Given basic functionality is currently enabled
And user is in Settings > Security & Privacy > Basic Functionality

When user toggles "Basic Functionality" to OFF
Then confirmation modal appears
When user confirms the action
Then Redux state updates to basicFunctionalityEnabled: false  
And MultichainAccountService.setBasicFunctionality is called with enabled: false
And console shows "MultichainAccountService: Setting basic functionality disabled"
And EVM provider remains active
And non-EVM providers (Solana, etc.) are disabled

---

Scenario: user enables basic functionality via backup sync flow
Given basic functionality is disabled
And user is in backup & sync confirmation modal

When user confirms "Turn On" 
Then MultichainAccountService.setBasicFunctionality is called with enabled: true
And providers are enabled and alignment is triggered

## **Screenshots/Recordings**

<img width="1127" height="1011" alt="image" src="https://github.com/user-attachments/assets/f5be2102-1b18-4b70-aee6-f36223d1be82" />


## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [ ] I've included tests if applicable
- [x] I've documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I've applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

[MUL-343]: https://consensyssoftware.atlassian.net/browse/MUL-343?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
feat: trigger MultichainAccountService basic functionality updates

Commit:
refactor(multichain): simplify handler signature for setBasicFunctionality action

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
chore: bump multichain-account-service version to preview release

Commit:
chore: bump multichain-account-service version to preview release

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

# Conflicts:
#	package.json
#	yarn.lock

Commit:
feat: trigger multichain provider state alignment

- Remove debug console.log statements to comply with ESLint rules
- Call MultichainAccountService.setBasicFunctionality() to trigger provider state alignment when toggling basic functionality

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
improve test coverage

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
fix unit test snapshot

Commit:
feat(multichain): add MultichainAccountService integration to BasicFunctionalityModal

- Add Engine.context.MultichainAccountService.setBasicFunctionality call when toggling basic functionality
- Include comprehensive test coverage for the new integration
- Handle errors gracefully with console logging
- Ensure alignment triggers when basic functionality is enabled

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
test: update BasicFunctionalityModal snapshots

- Update snapshot to reflect MultichainAccountService integration
- Remove obsolete snapshots from previous test structure

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

ccharly commented on package.json:
```diff
@@ -257,7 +257,7 @@
     "@metamask/mobile-wallet-protocol-core": "^0.0.6",
     "@metamask/mobile-wallet-protocol-wallet-client": "^0.0.6",
-    "@metamask/multichain-account-service": "^0.4.0",
+    "@metamask/multichain-account-service": "npm:@metamask-previews/multichain-account-service@0.5.0-preview-970969ca",
```
We should use the `0.5.0` now.

ccharly commented on app/components/UI/BasicFunctionality/BasicFunctionalityModal/BasicFunctionalityModal.tsx:
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable */
```
Yup, we need to remove this one.

ccharly commented on app/components/UI/BasicFunctionality/BasicFunctionalityModal/BasicFunctionalityModal.tsx:
```diff
@@ -67,11 +69,26 @@ const BasicFunctionalityModal = ({ route }: Props) => {
+    bottomSheetRef.current?.onCloseBottomSheet(async () => {
+      const newBasicFunctionalityState = !isEnabled;
+      dispatch(toggleBasicFunctionality(newBasicFunctionalityState));
+      ///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
```
Those code-fence looked inappropriate? 🤔

ccharly commented on app/components/UI/Identity/ConfirmTurnOnBackupAndSyncModal/ConfirmTurnOnBackupAndSyncModal.tsx:
```diff
@@ -16,6 +16,9 @@ import { toggleBasicFunctionality } from '../../../../actions/settings';
 import { useParams } from '../../../../util/navigation/navUtils';
 import { ConfirmTurnOnBackupAndSyncModalNavigateParams } from '../BackupAndSyncToggle/BackupAndSyncToggle';
 import { InteractionManager } from 'react-native';
+///: BEGIN:ONLY_INCLUDE_IF(keyring-snaps)
```
Same here.

Commit:
feat: implement BasicFunctionality toggle integration with multichain account alignment

- Add basic functionality toggle listener in settings actions
- Integrate with MultichainWalletService for account alignment
- Update modal component with proper ESLint rule usage
- Fix deprecated @typescript-eslint rule reference

Commit:
yarn install

Commit:
refactor: remove obsolete ESLint rule disabling in BasicFunctionalityModal component

Commit:
fix: await MultichainAccountService call to prevent race condition

- Add proper await to setBasicFunctionality call in thunk
- Change to try-catch pattern for better async error handling
- Resolves cursor bot async race condition warning

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
fix: separate plain action from thunk for store initialization

- Create setBasicFunctionality plain action for store initialization
- Keep toggleBasicFunctionality thunk for user interactions
- Fixes TypeScript error where store dispatch expected plain action
- Resolves race condition and parameter issues flagged by cursor bot

Commit:
fix: update ConfirmTurnOnBackupAndSyncModal test for thunk action

- Mock toggleBasicFunctionality as thunk action
- Update test assertion to check function call instead of dispatch comparison
- Test now passes correctly with thunk implementation

Commit:
fix: update toggleBasicFunctionality mock to return thunk function

- Fixed mock in ConfirmTurnOnBackupAndSyncModal.test.tsx to properly simulate thunk action creator
- Mock now returns jest.fn(() => jest.fn()) instead of jest.fn() to match BasicFunctionalityModal.test.js
- Addresses cursor bot feedback about undefined return value

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
bump multichain-account-service to ver 0.6.0

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

Commit:
refactor: remove await from MultichainAccountService call for better UX

- Remove await to prevent blocking user during potentially long alignment operations
- Alignment now runs in background while toggle responds immediately
- Add TODO comment to revisit when wallet state/status UI is available
- Maintain error handling with promise .catch() plus outer try-catch

Addresses reviewer feedback from Charly about UX performance concerns

Commit:
simplified the error handling for setting basic functionality in MultichainAccountService by removing the outer try-catch block

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo commented on app/actions/settings/index.js:
```diff
@@ -54,13 +54,35 @@ export function setUseBlockieIcon(useBlockieIcon) {
+    // Call MultichainAccountService to update provider states and trigger alignment
+    const Engine = require('../../core/Engine').default;
+    try {
+      await Engine.context.MultichainAccountService.setBasicFunctionality(
```
I totally agree. We don't want mobile to be blocked, hence we shouldn't await. In extension we can because... the dispatched action would run in background, right?

ccharly replied:
I think yes. And again, once we have the "wallet state/status" available, we will have a way to show a different UI to improve user feedback on the on-going alignment of each wallets!

Commit:
fix var naming and unit test

Commit:
try to fix broken unit test

Commit:
fix lint issue

ccharly commented on app/actions/settings/index.js:
```diff
@@ -55,9 +55,23 @@ export function setUseBlockieIcon(useBlockieIcon) {
+        'Failed to set basic functionality on MultichainAccountService:',
+        error,
+      );
+    });
```
This is done on purpose, to avoid blocking the whole action.

We will have another way to report the wallet status change (triggered once basic functionality is toggled).

ccharly reviewed (approved):
LGTM + Tested!

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

fabiobozzo committed:
Merge branch 'main' into feat/MUL-343-trigger-multichain-alignment

mathieuartu reviewed (approved):
LGTM!
